### DISCUSSION TITLE
Merge two model eloquent and using Uinion

### DISCUSSION THREAD

**Discussion by saeed-rostami** [nuwave/lighthouse#2555] [General]
hello
on my app i have two Model eloquent , Podcast and Episode...i want to make query that based on my logic, merge these models and return the data...(mix of podcast and episode)...

```
union Posts= Podcast | Episode

type Podcast {
  id: ID!
}

type Episode{
  employeeId: ID!
} 
```

my question is how should be my type Query  ??

type Query {
posts : [?] ? 
}

**Comment by spawnia** [comment]
1. Use singular for union types. Otherwise, it is hard to express having more than one result.
2. Always use `id: ID!` as the unique identifier for object types. Makes it easier to handle in the client, some libraries such as Apollo depend on this for caching.
3. You can use your union type as the result type, as in `posts: [Post!]!`. You will need a custom resolver and maybe a type resolver.

**Comment by saeed-rostami** [comment]
Thank you for your response...I solved the first and second cases that you recommended.  

I made a class named Feed and a schema named feed.graphql .

my code in Feed final class : 

```
<?php

namespace App\GraphQL\Queries;

use Modules\Castalk\Entities\Podcast;
use Modules\Castalk\Entities\PodcastEpisode;

final class Feed
{
    /**
     * @param  null  $_
     * @param  array{}  $args
     */
    public function slider($_, array $args)
    {
        $podcasts = Podcast::query()
        ->get();

        $episodes = PodcastEpisode::query()
        ->get();

        return $podcasts->toBase()->merge($episodes);
    }
}
```

and this my code in feed.graphql

```
extend type Query {
    slider: [Post]  @field(resolver: "App\\GraphQL\\Queries\\Feed@slider")
}

union Post = Podcast | PodcastEpisode
```


it works properly as i want but do you think this is right approach? or there are other best practice?


my next question is it possible to paginate the result? i used @paginate directive and error : 

  **"debugMessage": "Could not determine a model name for the '@paginate' directive on 'slider.",**

thanks in advanced

**Reply by spawnia** [reply]
I recommend using the convention-over-configuration class based default resolver classes, as described in https://lighthouse-php.com/master/the-basics/fields.html#resolver-precedence. So in your case, the class should be named `Slider` and define a method `__invoke`.

The code can be more elegant:

```php
return [
    ...Podcast::all(),
    ...PodcastEpisode::all(),
];
```

Pagination is generally possible, using `@paginate` with the custom `resolver` option, see https://lighthouse-php.com/master/api-reference/directives.html#paginate `Custom resolver`.

**Comment by saeed-rostami** [comment]
Thanks lot my friend, it works... 
actually i am new in grapqhl and lighthouse-php, the most parts of the document is fluent and comprehensive but  CUSTOM DIRECTIVES  section is very brief and i hardly know about them,  for instance the diference between TypeManipulator,  TypeMiddleware, TypeResolver and same as in fields are very short explained, It would be great if a simple example was provided for each of these parts...
Thank you for your efforts

**Reply by spawnia** [reply]
Given there are a lot of use cases for custom directives, the documentation does not provide examples for most interfaces.
It is advised to look at the Lighthouse source code to find directives that implement certain interfaces to learn more.
